react-native/react/actions/login2.js

```diff
@@ -277,25 +163,302 @@ export function logout () {
+
+    engine.rpc('device.deviceAdd', {}, incomingMap, (error, response) => {
+      console.log(error)
+    })
+  }
+}
+
+function makeKex2IncomingMap (dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle, successType) {
+  return {
+    'keybase.1.provisionUi.chooseProvisioningMethod': (param, response) => {
+      response.result(provisionMethod)
+    },
+    // TODO remove this when KEX2 supports not asking for this
+    'keybase.1.loginUi.getEmailOrUsername': (param, response) => {
+      const { username } = getState().login2.userPass
+      if (!username) {
+        dispatch(askForUserPass(userPassTitle, userPassSubtitle, () => {
+          const { username } = getState().login2.userPass
+          response.result(username)
+        }))
+      } else {
+        response.result(username)
+      }
+    },
+    'keybase.1.secretUi.getPaperKeyPassphrase': (param, response) => {
+      dispatch(askForPaperKey((paperKey) => {
+        response.result(paperKey)
+      }))
+    },
+    'keybase.1.secretUi.getKeybasePassphrase': (param, response) => {
+      const { passphrase } = getState().login2.userPass
+      if (!passphrase) {
+        dispatch(askForUserPass(userPassTitle, userPassSubtitle, () => {
+          const { passphrase } = getState().login2.userPass
+          response.result(passphrase)
+        }))
+      } else {
+        response.result(passphrase)
+      }
+    },
+    'keybase.1.secretUi.getSecret': (param, response) => {
+      const { passphrase } = getState().login2.userPass
+      if (!passphrase) {
+        dispatch(askForUserPass(userPassTitle, userPassSubtitle, () => {
+          const { passphrase } = getState().login2.userPass
+          response.result({
+            text: passphrase,
+            canceled: false,
+            storeSecret: true
+          })
+        }))
+      } else {
+        response.result({
+          text: passphrase,
+          canceled: false,
+          storeSecret: true
+        })
+      }
+    },
+    'keybase.1.provisionUi.PromptNewDeviceName': (param, response) => {
+      const { existingDevices } = param
+      dispatch(askForDeviceName(existingDevices, () => {
+        const { deviceName } = getState().login2.deviceName
+        response.result(deviceName)
+      }))
+    },
+    'keybase.1.logUi.log': (param, response) => {
+      console.log(param)
+      response.result()
+    },
+    'keybase.1.provisionUi.ProvisioneeSuccess': (param, response) => {
+      response.result()
+    },
+    'keybase.1.provisionUi.ProvisionerSuccess': (param, response) => {
+      response.result()
+      const uri = getCurrentURI(getState()).last()
+
+      const onDevicesTab = getCurrentTab(getState()) === DEVICES_TAB
+      const onCodePage = uri && uri.getIn(['parseRoute']) &&
+        uri.getIn(['parseRoute']).componentAtTop && uri.getIn(['parseRoute']).componentAtTop.component === CodePage
```

Comment:
might be nice to move this into the router 'isTopComponent(CodePage)' or something